crypto/rand/drbg_lib.c

```diff
@@ -410,34 +595,43 @@ static int drbg_bytes(unsigned char *out, int count)
     return ret;
 }
 
+/* Implements the default OpenSSL RAND_add() method */
 static int drbg_add(const void *buf, int num, double randomness)
 {
-    unsigned char *in = (unsigned char *)buf;
-    unsigned char *out, *end;
+    int ret = 0;
+    RAND_DRBG *drbg = RAND_DRBG_get0_global();
 
-    CRYPTO_THREAD_write_lock(rand_bytes.lock);
-    out = &rand_bytes.buff[rand_bytes.curr];
-    end = &rand_bytes.buff[rand_bytes.size];
+    if (drbg == NULL)
+        return 0;
 
-    /* Copy whatever fits into the end of the buffer. */
-    for ( ; --num >= 0 && out < end; rand_bytes.curr++)
-        *out++ = *in++;
+    if (num < 0 || randomness < 0.0)
```

Comment:
Would an additional check of `randomess < UINT_MAX / 8` be worthwhile here?

It is possible for either:
1. a _double_ to have a value too large to fit a _unsigned int_ or
1. for a _unsigned int_ to not convert to a _double_ without rounding away from zero.

Later code casts to an _unsigned int_ even though the variable used is _size_t_.  Hence the check against_UINT_MAX_ not _SIZE_MAX_.

Reply:
`randomness` is already bounded by `drbg->max_entropylen`, which is a  small fixed value. Doesn't this already safe my day?

```C
    if (randomness > (double)drbg->max_entropylen) {
        /*
         * The purpose of this check is to bound |randomness| by a
         * relatively small value in order to prevent an integer
         * overflow when multiplying by 8 in the rand_drbg_restart()
         * call below.
         */
        return 0;
    }
```

As for the casts: The first `(unsigned int)` cast for `num` was on purpose, because I did not want to worry about sign extension. The second cast should better be a `(size_t)` cast. I changed the call as follows:

```C
    ret = rand_drbg_restart(drbg, buf,
                            (size_t)(unsigned int)num,
                            (size_t)(8*randomness));
```